File: 240807/우리는 하나/we-are-the-one.py
n, k, u, d = tuple(map(int, input().split()))
arr = [
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [
    [0 for _ in range(n)]
    for _ in range(n)
]

arr_num = []

for i in range(n):
    for j in range(n):
        arr_num.append((i, j))



def in_range(x, y):
    return x>=0 and x<n and y>=0 and y<n

def can_go(x1, y1, x2, y2):


    if not in_range(x2, y2):
        return False
    diff = abs(arr[x1][y1] - arr[x2][y2])

    if visited[x2][y2] == 1 or not (diff >= u and diff <= d):
        return False
    return True

from collections import deque
q = deque()

def print_visited():
    for i in range(n):
        for j in range(n):
            print(visited[i][j], end=" ")
        print()

def bfs():
    dxs, dys = [0, 1, 0, -1], [1, 0, -1, 0]

    while q:
        
        x, y = q.popleft()

        for dx, dy in zip(dxs, dys):
            nx, ny = x+dx, y+dy
            if can_go(x, y, nx, ny):
                # print(x, y, nx, ny)
                visited[nx][ny] = 1
                q.append((nx, ny))

        # print_visited()

ans = []

def reset_visited():
    for i in range(n):
        for j in range(n):
            visited[i][j] = 0


result = 0 
def collect_and_bfs():
    start_points = []
    reset_visited()

    for point in ans:
        start_points.append(arr_num[point])
    
    # print(start_points)
    
    # 그러면 0,0과 0,1이 저장 됨 
    for point in start_points:
        x, y = point 
        visited[x][y] = 1
        q.append((x, y))
        bfs()

    count = 0 

    for i in range(n):
        for j in range(n):
            if visited[i][j] == 1:
                count+=1 

    return count

def choose(num, cnt):
    global result
    if cnt == k:
        # print(ans)
        result = max(result, collect_and_bfs())
        return

    if num == (n*n):
        return
    
    ans.append(num)
    choose(num+1, cnt+1)
    ans.pop()

    choose(num+1, cnt)

choose(0, 0)
print(result)

File: 241004/우리는 하나/we-are-the-one.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <tuple>
#include <cmath>

#define MAX_N 100
#define DIR_N 4

using namespace std;

int grid[MAX_N][MAX_N];
bool visited[MAX_N][MAX_N];

vector<pair<int, int>> city; 
vector<pair<int, int>> selected;

queue<pair<int, int>> q; 

int n, k, u, d;

bool in_range(int x, int y){
    return x>=0 && x<n && y>=0 && y<n;
}

void reset_visited(){
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            visited[i][j] = false; 
}

void vector_to_q(){
    for (auto s : selected){
        q.push(s);
        visited[s.first][s.second] = true;
    }
}

bool can_go(int x, int y, int h1, int h2){
    if (!in_range(x, y))
        return false;

    int diff = abs(h1 - h2);
    
    if (visited[x][y] == true)
        return false;
    
    if (diff >= u && diff <= d)
        return true;
    
    return false;
}


void print_selected(){
    for(auto s : selected){
        cout << s.first << " " << s.second ;
    cout << endl; 
    }
}

int count_visited(){
    int count = 0;
    for (int i = 0; i < n; i++){
        for (int j = 0; j < n; j++){
            if (visited[i][j] == true)
                count += 1;
        }
    }
    return count; 

}

void bfs(){
    int dx[DIR_N] = {0, 1, 0, -1};
    int dy[DIR_N] = {1, 0, -1, 0};

    while(!q.empty()){
        int x, y;
        x = q.front().first;
        y = q.front().second;
        q.pop();

        for(int i = 0; i<DIR_N; i++){
            int nx = x+dx[i];
            int ny = y+dy[i];

            if(can_go(nx, ny, grid[x][y], grid[nx][ny])){
                q.push(make_pair(nx, ny));
                visited[nx][ny] = true;
            }
        }


    }
}
void print_visited(){
    for (int i = 0; i < n; i++){
        for (int j = 0; j < n; j++){
            cout << visited[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl; 
}

int result = 0; 

void choose(int idx, int cnt){
    if (cnt == k){
        reset_visited();
        vector_to_q();
        bfs();
        // print_selected();
        // print_visited();
        
        result = max(result, count_visited());

        // cout << result << endl; 
        return;
    }

    if (idx >= n*n)
        return;

    selected.push_back(city[idx]);
    choose(idx+1, cnt+1);
    selected.pop_back();

    choose(idx+1, cnt);

}



int main() {
    // 여기에 코드를 작성해주세요.

    cin >> n >> k >> u >> d;

    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            cin >> grid[i][j];
            city.push_back(make_pair(i, j));
            visited[i][j] = false;
        }
    }

    choose(0, 0);

    cout << result; 
    return 0;
}